src/sample.rs

```diff
@@ -0,0 +1,329 @@
+
+    fn sample<R: Rng>(&self, rng: &mut R, amount: usize) -> Vec<T> {
+        self.as_slice().sample(rng, amount)
+    }
+}
+
+impl<'a, T> SampleRef for &'a [T] {
+    type SampledRef = Vec<&'a T>;
+
+    fn sample_ref<R: Rng>(&self, rng: &mut R, amount: usize) -> Vec<&'a T> {
+        let indices = sample_indices(rng, self.len(), amount);
+
+        let mut out = Vec::with_capacity(amount);
+        out.extend(indices.iter().map(|i| &self[*i]));
+        out
+    }
+}
+
+// TODO: It looks like implementing this depends on RFC 1598 being implemented.
+// See this: https://github.com/rust-lang/rfcs/issues/1965
+//
+// impl<'a, T> SampleRef for Vec<&'a T>{
+//     type SampledRef = Vec<&'a T>;
+//
+//     fn sample_ref<R: Rng>(&'a self, rng: &mut R, amount: usize) -> Vec<&'a T> {
+//        self.as_slice().sample_ref(rng, amount)
+//     }
+// }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use {thread_rng, XorShiftRng, SeedableRng};
+
+    #[test]
+    fn test_sample_reservoir() {
+        let min_val = 1;
+        let max_val = 100;
+
+        let mut r = thread_rng();
+        let vals = (min_val..max_val).collect::<Vec<i32>>();
+        let small_sample = sample_reservoir(&mut r, vals.iter(), 5);
+        let large_sample = sample_reservoir(&mut r, vals.iter(), vals.len() + 5);
+
+        assert_eq!(small_sample.len(), 5);
+        assert_eq!(large_sample.len(), vals.len());
+
+        assert!(small_sample.iter().all(|e| {
+            **e >= min_val && **e <= max_val
+        }));
+    }
+
+    #[test]
+    /// This test mainly works by asserting that the two cases are equivalent,
+    /// as well as equivalent to the exported function.
+    fn test_sample_indices() {
+        let xor_rng = XorShiftRng::from_seed;
+
+        let max_range = 100;
+        let mut r = thread_rng();
+
+        for length in 1usize..max_range {
+            let amount = r.gen_range(0, length);
+            let seed: [u32; 4] = [
+                r.next_u32(), r.next_u32(), r.next_u32(), r.next_u32()
+            ];
+
+            println!("Selecting indices: len={}, amount={}, seed={:?}", length, amount, seed);
+
+            // assert that the two methods give exactly the same result
+            let inplace = sample_indices_inplace(
+                &mut xor_rng(seed), length, amount);
+            let cache = sample_indices_cache(
+                &mut xor_rng(seed), length, amount);
+            assert_eq!(inplace, cache);
+
+            // assert the basics work
+            let regular = sample_indices(
+                &mut xor_rng(seed), length, amount);
+            assert_eq!(regular.len(), amount);
+            assert!(regular.iter().all(|e| *e < length));
+            assert_eq!(regular, inplace);
+
+            // just for fun, also test sampling from a vector
+            let vec: Vec<usize> = (0..length).collect();
+            assert_eq!(
+                vec.sample(&mut xor_rng(seed), amount),
+                regular,
```

Comment:
Trailing comma may be an issue in some builds? At least, Travis complains on some builds (also line 325)